### ISSUE TITLE
Keyword "Are" is not giving correct results

### ISSUE THREAD

**Issue by Laurira** [keeps/db-visualization-toolkit#116]
Description: 
For example we have a database where we have persons' names in it. I can search other names but for example "Are" (that is very popular name in Estonia) is giving too many results.

Browser (Internet explorer, Chrome, Firefox, etc):

Browser version:

Steps required to reproduce the bug:

The issue can also be reproduced in your database "World".
1. Open visualisation toolkit -> database "World"
2. Click "Data" -> "Country"
3. Click the arrow near the search button then the form opens where you can see the fields.
4. Insert "Estonia" as the "Name" and click "Search"
You will be displayed 1 result.
5. Insert "Are" as the "Name" and click "Search"
You will be displayed 239 results.

While you fix it then please try to understand if there are more such keywords that can not be used?

![image](https://user-images.githubusercontent.com/3479156/31992822-0a9bd698-b984-11e7-878e-f7d53b7a03a0.png)

**Comment by Laurira** [comment]
Keywords "is", "or", "and", "for", "if" ... are also not working correctly. Probably there are a lot of more strings that you cannot use.

**Comment by chalkos** [comment]
I was writing a comment to ask you to try more stopwords :)

Is 239 the total number of records?

**Comment by Laurira** [comment]
Yes, 239 is the total number of records.

**Comment by chalkos** [comment]
A search containing only stopwords is being treated as an empty search and is not filtering anything out (that is why it shows all records in the search results). The correct behaviour would be to not show any results.

**Comment by jmaferreira** [comment]
Or.. remove stop words altogether as they are language specific...

**Comment by chalkos** [comment]
Stopwords make a lot of sense when it comes to (accurately, efficiently) finding data and reducing the number of "false matches" in search results.

Apart from the bug (the way the search is putting all the parameters together), the best way to solve the stopwords issue is to allow modifying the config file to use stopwords from another language (or even use stopwords from multiple languages).

Another solution is to index data with and without stopwords, and start by showing the search results that match even when filtering out stopwords (by ranking them higher). Idea from: https://stackoverflow.com/a/31850227

These 3 solutions (bug fix, modifying config, change indexing and use ranked searches) are not exclusive. In fact they complement each other to provide a good balance between the number of search results and the search results actually showing you what you are looking for.

**Comment by jmaferreira** [comment]
I agree…
Let’s start with the fix that imposes the smaller effort and see how it goes.

PS: Beware of the sentence 'reducing the number of "false matches”’. In fact, removing stop words reduces precision and increases recall, with in fact increases false matches...

**Comment by chalkos** [comment]
> Beware of the sentence 'reducing the number of "false matches”’. In fact, removing stop words reduces precision and increases recall, with in fact increases false matches...

Correct. I wanted to keep the recall but continue to display all records that the user might want to be looking for. See #119, where I hope I explained this better.

@laurira I divided this issue in 3 different issues:

* #117 fixes part of the problem by not showing any results when the search contains only stopwords (the results show all records, when none should be showing because, since it uses english stopwords, the search matches nothing)
* #118 is a rough fix for your problem, as it allows the customization of stopword lists, allowing you to set DBVTK to use use Estonian stopwords and thus showing you better search results for Estonian databases.
* #119 completes the circle for all cases, as it contains a fallback mechanism that does not exclude results that might be relevant. It is a solution that shows the records you are most likely looking for in the beginning of the results list, and also shows all the other matching records (ignoring stopword handling) after that (since you might want one of those in case you are searching in another language or searching in a specific field in which stopwords make little sense, like names).
